### PR TITLE
{toolchain} intel/2023.07

### PR DESCRIPTION
**Pull request by SebastianAchilles**
(created using `eb --new-pr`)

### CODE REVIEW

File: easybuild/easyconfigs/i/intel-compilers/intel-compilers-2023.2.1.eb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    },
+]
+checksums = [
+    {'l_dpcpp-cpp-compiler_p_2023.2.1.8_offline.sh':
+     'f5656b2f5bb5d904639e6ef1f90a2d2e760d2906e82ebc0dd387709738ca714b'},
+    {'l_fortran-compiler_p_2023.2.1.8_offline.sh':
+     'd4e36abc014c184698fec318a127f15a696b5333b3b0282aba1968b351207185'},
+]
+
+local_gccver = '12.3.0'
```
Comment (SebastianAchilles) [inline easybuild/easyconfigs/i/intel-compilers/intel-compilers-2023.2.1.eb:31]: Not sure if we should go with 12.3.0 or 13.2.0.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in total)
zen2-ubuntu-2304 - Linux Ubuntu 23.04 (Lunar Lobster), x86_64, AMD EPYC 7452 32-Core Processor (zen2), Python 3.11.2
See https://gist.github.com/SebastianAchilles/1505e535182b414cff0964b63b1e17b4 for a full test report.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in total)
skl-centos-79 - Linux CentOS Linux 7.9.2009, x86_64, Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz (skylake), Python 3.6.8
See https://gist.github.com/SebastianAchilles/5e2800fae2e8f432d5964a703cd47994 for a full test report.

**Comment by SebastianAchilles** [comment]
@boegelbot please test @ generoso

**Comment by SebastianAchilles** [comment]
@boegelbot please test @ jsc-zen2

File: easybuild/easyconfigs/i/intel-compilers/intel-compilers-2023.2.1.eb (hunk 1 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/i/intel-compilers/intel-compilers-2023.2.1.eb:31]: I found the following in https://www.intel.com/content/www/us/en/developer/articles/release-notes/oneapi-c-compiler-release-notes.html:

> Note: Intel® C++ Compiler Classic doesn't support GCC 13 compatibility mode. For GCC 13 compatibility mode, please transition to Intel®  oneAPI DPC++/C++ Compiler.

**Comment by casparvl** [comment]
Test report by @casparvl
**FAILED**
Build succeeded for 4 out of 7 (7 easyconfigs in total)
tcn222.local.snellius.surf.nl - Linux RHEL 8.6, x86_64, AMD EPYC 7H12 64-Core Processor, Python 3.6.8
See https://gist.github.com/casparvl/cc08b2d35904d779acc1ab7b2bc486a2 for a full test report.

**Comment by boegel** [comment]
Hmm, seeing lots of `icc: remark #10441: The Intel(R) C++ Compiler Classic (ICC) is deprecated` messages, why/how is `icc` still being used for `imkl-FFTW`?

**Comment by casparvl** [comment]
The failure seems to be related to our site using EB's RPATH support. The reason this leads to errors is because of a combination of `-Wall -Werror`, that turns on all warnings, and then turns all warnings into errors. Ooof, that is very conservative. The EasyBuild RPATH wrappers inject `-Wl,rpath` arguments, which are unused on pure compilation steps (they are only used when linking). Hence the error.

The _solution_ would be to make sure that either warnings are not enabled for unused arguments _or_ that they are not turned into errors. I'll have a look to see if any prior versions of FFTW also did this and what we did to avoid this. I remember having seen this error pattern before, but can't remember with which software that was...

**Comment by casparvl** [comment]
Found it: used to have this issue with Clang, see [this code section](https://github.com/easybuilders/easybuild-easyblocks/blob/develop/easybuild/easyblocks/c/clang.py#L503) in the clang EasyBlock.

Maybe we can do something similar for FFTW? I _do_ wonder why we haven't see this with older versions, but maybe the `-Wall` or `-Werror` is new to this version...?

**Comment by casparvl** [comment]
I tried a somewhat naive addition of
```
    def build_step(self):
        """Build imkl-FFTW"""
        # Avoid unused-command-line-argument error when builindg with RPATH support
        # See https://github.com/easybuilders/easybuild-easyconfigs/pull/18439#issuecomment-1662671054
        if build_option('rpath'):
            cflags = os.getenv('CFLAGS')
            cxxflags = os.getenv('CXXFLAGS')
            setvar('CFLAGS', "%s %s" % (cflags, '-Wno-unused-command-line-argument'))
            setvar('CXXFLAGS', "%s %s" % (cxxflags, '-Wno-unused-command-line-argument'))
        super(EB_imkl_minus_FFTW, self).build_step()
```

to the EasyBlock. That doesn't help. Looking more closely, I see that for a lot of build targets, we don't actually see the problem:

```
== 2023-08-04 14:04:04,850 run.py:648 INFO cmd "make -f makefile libintel64 compiler=intel install_to=$INSTALL_DIR  " exited with exit code 0 and output:
make -f  makefile lib _IA=intel64
make[1]: Entering directory '/gpfs/nvme1/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/interfaces/fftw3xf'
mkdir -p /scratch-nvme/1/casparl/ebbuildpath/imklFFTW/2023.2.0/iimpi-2023.07/tmp6ad7w836/obj_intel
icc  -Wall -Werror -fPIC -I/scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/include -I/scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/include/fftw     -c /scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/interfaces/fftw3xf/wrappers/fftw_alignment_of.c -o /scratch-nvme/1/casparl/ebbuildpath/imklFFTW/2023.2.0/iimpi-2023.07/tmp6ad7w836/obj_intel/fftw_alignment_of.o
icc: remark #10441: The Intel(R) C++ Compiler Classic (ICC) is deprecated and will be removed from product release in the second half of 2023. The Intel(R) oneAPI DPC++/C++ Compiler (ICX) is the recommended compiler moving forward. Please transition to use this compiler. Use '-diag-disable=10441' to disable this message.
icc  -Wall -Werror -fPIC -I/scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/include -I/scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/include/fftw     -c /scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/interfaces/fftw3xf/wrappers/fftw_alloc_complex.c -o /scratch-nvme/1/casparl/ebbuildpath/imklFFTW/2023.2.0/iimpi-2023.07/tmp6ad7w836/obj_intel/fftw_alloc_complex.o
...
```
even though `-Wall -Werror` are also passed here. I've also noted that CFLAGS get reset _right_ before building this target:

```
== 2023-08-04 14:03:05,015 environment.py:93 INFO Environment variable CFLAGS set to -fPIC (previous value: '')
== 2023-08-04 14:03:05,015 imkl.py:275 INFO Changed to interface fftw3xf directory /scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/interfaces/fftw3xf
== 2023-08-04 14:03:05,015 run.py:236 INFO running cmd: make -f makefile libintel64 compiler=intel install_to=$INSTALL_DIR
== 2023-08-04 14:04:04,850 run.py:648 INFO cmd "make -f makefile libintel64 compiler=intel install_to=$INSTALL_DIR  " exited with exit code 0 and output:
make -f  makefile lib _IA=intel64
```
So my addition of `-Wno-unused-command-line-argument` isn't even passed to the compiler there.

The problematic section is this:

```
== 2023-08-04 14:04:04,879 environment.py:93 INFO Environment variable SPEC_OPT set to  (previous value: '-fPIC')
== 2023-08-04 14:04:04,879 environment.py:93 INFO Environment variable COPTS set to  (previous value: '-fPIC')
== 2023-08-04 14:04:04,879 environment.py:93 INFO Environment variable CFLAGS set to  (previous value: '-fPIC')
== 2023-08-04 14:04:04,880 imkl.py:275 INFO Changed to interface fftw2x_cdft directory /scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/interfaces/fftw2x_cdft
== 2023-08-04 14:04:04,881 run.py:236 INFO running cmd: make -f makefile libintel64 compiler=intel interface=lp64 PRECISION=MKL_DOUBLE
== 2023-08-04 14:04:05,247 build_log.py:171 ERROR EasyBuild crashed with an error (at easybuild/base/exceptions.py:126 in __init__): cmd "make -f makefile libintel64 compiler=intel interface=lp64 PRECISION=MKL_DOUBLE" exited with exit code 2 and output:
make clean mkobjdir wraplib ILP_EXT=_lp64 _IA=intel64
make[1]: Entering directory '/gpfs/nvme1/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/interfaces/fftw2x_cdft'
rm -rf /scratch-nvme/1/casparl/ebbuildpath/imklFFTW/2023.2.0/iimpi-2023.07/tmpbg6vcw4z/obj_DOUBLE_intel64_lp64
rm -f /scratch-nvme/1/casparl/ebbuildpath/imklFFTW/2023.2.0/iimpi-2023.07/tmpbg6vcw4z/libfftw2x_cdft_DOUBLE_lp64.a
mkdir -p /scratch-nvme/1/casparl/ebbuildpath/imklFFTW/2023.2.0/iimpi-2023.07/tmpbg6vcw4z/obj_DOUBLE_intel64_lp64
mkdir -p /scratch-nvme/1/casparl/ebbuildpath/imklFFTW/2023.2.0/iimpi-2023.07/tmpbg6vcw4z
mpiicc -c  -Wall -Werror -std=c99 -Wall -Werror -std=c99    \
         -DMKL_DOUBLE -I../../include -I../../include/fftw wrappers/1d_create_plan.c -o /scratch-nvme/1/casparl/ebbuildpath/imklFFTW/2023.2.0/iimpi-2023.07/tmpbg6vcw4z/obj_DOUBLE_intel64_lp64/1d_create_plan.o
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/imkl-FFTW/2023.2.0-iimpi-2023.07/lib: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/imkl-FFTW/2023.2.0-iimpi-2023.07/lib64: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=$ORIGIN: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=$ORIGIN/../lib: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=$ORIGIN/../lib64: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,--disable-new-dtags: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/mkl/2023.2.0/lib/intel64: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/imkl/2023.2.0/compiler/2023.2.0/linux/compiler/lib/intel64_lin: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/impi/2021.10.0-intel-compilers-2023.2.1/mpi/2021.10.0/libfabric/lib: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/impi/2021.10.0-intel-compilers-2023.2.1/mpi/2021.10.0/lib/release: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/impi/2021.10.0-intel-compilers-2023.2.1/mpi/2021.10.0/lib: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/UCX/1.14.1-GCCcore-12.3.0/lib: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/numactl/2.0.16-GCCcore-12.3.0/lib: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/intel-compilers/2023.2.1/tbb/2021.10.0/lib/intel64/gcc4.8: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/intel-compilers/2023.2.1/compiler/2023.2.1/linux/compiler/lib/intel64_lin: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/intel-compilers/2023.2.1/compiler/2023.2.1/linux/lib/x64: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/intel-compilers/2023.2.1/compiler/2023.2.1/linux/lib: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/binutils/2.40-GCCcore-12.3.0/lib: 'linker' input unused [-Werror,-Wunused-command-line-argument]
icx: error: -Wl,-rpath=/scratch-nvme/1/casparl/generic/software/zlib/1.2.13-GCCcore-12.3.0/lib: 'linker' input unused [-Werror,-Wunused-command-line-argument]
make[1]: *** [makefile:203: /scratch-nvme/1/casparl/ebbuildpath/imklFFTW/2023.2.0/iimpi-2023.07/tmpbg6vcw4z/obj_DOUBLE_intel64_lp64/1d_create_plan.o] Error 1
```
So, here it is `mpiicc` that is invoked, which then invokes `icx`. Somehow, then, and only then, does this combination of `-Wall and -Werror` seem to lead to the error.

Question now is: what do I need to change to modify the CFLAGS that are used _at this particular stage_ in the build. Which file is calling for that environment change? Well, that would be [here](https://github.com/easybuilders/easybuild-easyblocks/blob/d8a8fa30310cc7d69a670ea345915ab5dec224c4/easybuild/easyblocks/i/imkl.py#L271).

I'll try to modify `imkl.py` first on that line by hardcoding `flags + ' -Wunused-command-line-argument'` there. If that resolves it, I'll make it 'nice' by making it conditional on the `rpath` build option.

**Comment by casparvl** [comment]
Yep, putting
```
                env.setvar('CFLAGS', flags + ' -Wno-unused-command-line-argument')
```
in `imkl.py` works. I'll make this more pretty by making it conditional on the use of `-rpath`, and then make a PR to easyblocks.

**Comment by casparvl** [comment]
Test report by @casparvl
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2975
**FAILED**
Build succeeded for 0 out of 7 (7 easyconfigs in total)
tcn174.local.snellius.surf.nl - Linux RHEL 8.6, x86_64, AMD EPYC 7H12 64-Core Processor, Python 3.6.8
See https://gist.github.com/casparvl/dca543b61a9ba54296533e87d5f2d68b for a full test report.

**Comment by casparvl** [comment]
> Hmm, seeing lots of `icc: remark #10441: The Intel(R) C++ Compiler Classic (ICC) is deprecated` messages, why/how is `icc` still being used for `imkl-FFTW`?

So, this PR looks good to me in the sense that it is building now. I guess the only open question is this from @boegel . I'm not fully up to date on what we now do by default in EB: did we already switch to the dcpp compilers by default? I.e. is that why @boegel is surprised here about the use of `icc`?

**Comment by casparvl** [comment]
Test report by @casparvl
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2975
**FAILED**
Build succeeded for 0 out of 7 (7 easyconfigs in total)
tcn174.local.snellius.surf.nl - Linux RHEL 8.6, x86_64, AMD EPYC 7H12 64-Core Processor, Python 3.6.8
See https://gist.github.com/casparvl/636d5050b7fe1018040f3b04c2b35453 for a full test report.

**Comment by casparvl** [comment]
Discussed with @boegel : we should open a separate issue for the icc stuff. He had a short look, it seems the `makefile` for FFTW wrapper doesn't recognize `icx`. It _does_ recognize `mpiicc`, which then calls `icx` (cause the `rpath` error I head earlier, because `icx` is a clang-based compiler). That makes sense: I only saw this error in the log once a target was build with `mpiicc`, all targets build with `icc` completed fine.

**Comment by casparvl** [comment]
Test report by @casparvl
Using easyblocks from PR(s) https://github.com/easybuilders/easybuild-easyblocks/pull/2975
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in total)
tcn174.local.snellius.surf.nl - Linux RHEL 8.6, x86_64, AMD EPYC 7H12 64-Core Processor, Python 3.6.8
See https://gist.github.com/casparvl/69b710fe4c823b5527447e766bc33899 for a full test report.

**Comment by SebastianAchilles** [comment]
> @SebastianAchilles The plan here is still to change to `GCCcore/13.2.0` as base compiler for this, right?

Yes, that is the plan. Since a UCX for `GCCcore/13.2.0` was added in https://github.com/easybuilders/easybuild-easyconfigs/pull/18710 this PR can be updated now.

File: easybuild/easyconfigs/i/intel-compilers/intel-compilers-2023.2.1.eb (hunk 1 above)
Comment (SebastianAchilles) [reply easybuild/easyconfigs/i/intel-compilers/intel-compilers-2023.2.1.eb]: Done in https://github.com/easybuilders/easybuild-easyconfigs/pull/18439/commits/ccaeebc51b3d523d55a206aab7449c5b3380f410

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in total)
bwd-rockylinux-92 - Linux Rocky Linux 9.2 (Blue Onyx), x86_64, Intel(R) Core(TM) i7-6900K CPU @ 3.20GHz (broadwell), 2 x NVIDIA NVIDIA GeForce GTX 1060 6GB, 535.104.05, Python 3.9.16
See https://gist.github.com/SebastianAchilles/585d7ca3e2cf1c2c239b7d5948232d20 for a full test report.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in total)
skl-centos-79 - Linux CentOS Linux 7.9.2009, x86_64, Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz (skylake), Python 3.6.8
See https://gist.github.com/SebastianAchilles/741189ae9d3e023bb640c7626143928d for a full test report.

**Comment by SebastianAchilles** [comment]
@boegelbot please test @ jsc-zen2

**Comment by SebastianAchilles** [comment]
@boegelbot please test @ generoso

File: easybuild/easyconfigs/i/impi/impi-2021.10.0-intel-compilers-2023.2.1.eb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+description = "Intel MPI Library, compatible with MPICH ABI"
+
+toolchain = {'name': 'intel-compilers', 'version': '2023.2.1'}
+
+# see https://software.intel.com/content/www/us/en/develop/articles/oneapi-standalone-components.html
+source_urls = ['https://registrationcenter-download.intel.com/akdlm/IRC_NAS/4f5871da-0533-4f62-b563-905edfb2e9b7/']
+sources = ['l_mpi_oneapi_p_%(version)s.49374_offline.sh']
+checksums = ['ab2e97d87b139201a2e7dab9a61ac6e8927b7783b459358c4ad69a1b1c064f40']
+
+dependencies = [('UCX', '1.15.0-rc4')]
```
Comment (boegel) [inline easybuild/easyconfigs/i/impi/impi-2021.10.0-intel-compilers-2023.2.1.eb:14]: Hmm, not sure it's wise to use a release candidate, we should stick to 1.14.1 instead, no?

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/i/impi/impi-2021.10.0-intel-compilers-2023.2.1.eb (hunk 2 above)
Comment (boegel) [reply easybuild/easyconfigs/i/impi/impi-2021.10.0-intel-compilers-2023.2.1.eb:14]: Should change to 1.14.1 once #18877 is merged

```suggestion
dependencies = [('UCX', '1.14.1')]

**Comment by SebastianAchilles** [comment]
@boegelbot please test @ jsc-zen2

**Comment by SebastianAchilles** [comment]
@boegelbot please test @ generoso

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in total)
node3157.skitty.os - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz (skylake_avx512), Python 3.6.8
See https://gist.github.com/boegel/99bc2035add2ee9613de92e35d5e5254 for a full test report.

**Comment by boegel** [comment]
Going in, thanks @SebastianAchilles!

**Pull request merged**